src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/Infrastructure/AsyncQueryExecutorSupplier.cs

```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Components.QuickGrid.Infrastructure;
+
+internal static class AsyncQueryExecutorSupplier
+{
+    // The primary goal with this is to ensure that:
+    //  - If you're using EF Core, then we resolve queries efficiently using its ToXyzAsync async extensions and don't
+    //    just fall back on the synchronous IQueryable ToXyz calls
+    //  - ... but without QuickGrid referencing Microsoft.EntityFramework directly. That's because it would bring in
+    //    heavy dependencies you may not be using (and relying on trimming isn't enough, as it's still desirable to have
+    //    heavy unused dependencies for Blazor Server).
+    //
+    // As a side-effect, we have an abstraction IAsyncQueryExecutor that developers could use to plug in their own
+    // mechanism for resolving async queries from other data sources than EF. It's not really a major goal to make this
+    // adapter generally useful beyond EF, but fine if people do have their own uses for it.
+
+    private static readonly ConcurrentDictionary<Type, bool> IsEntityFrameworkProviderTypeCache = new();
+
+    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2111",
+               Justification = "In the unlikely case a necessary interface is trimmed away and the error wasn't caught in developnent," +
+        " the developer only risks not noticing an oppurtunity to improve performance.")]
```

Comment:
```suggestion
    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2111",
               Justification = "The reflection is a best effort to warn developers about sync-over-async behavior which can cause thread pool starvation.")]
```

I agree the trim-unsafe reflection is worth doing because it's better to catch this issue sometimes rather than not at all, but I think saying this is only a performance concern is understating it. I'd think blocking a thread on every query from the grid could pretty easily make the app completely unresponsive for other users.

Reply:
The main use cases are [A] working with in-memory data (someArray.AsQueryable, when you already got the data from somewhere - this is especially common on WebAssembly) or [B] Entity Framework or [C] remote API calls. None of these cases will result in thread-blocking I/O, so in mainstream scenarios all is well.

As for the uncommon case of an IQueryable that isn't from EF and does thread-blocking I/O, that's not good but is it really our business to actually stop people from doing it? How would we even know it does thread-blocking I/O? Should other parts of the product detect if you're using thread-blocking I/O from an IQueryable and stop it? I'm not sure there's much we could or should do about this, but am definitely open to suggestions!